tetris.man
.TH TETRIS 6 "21 Feb 94, V3.1" "Georgia Tech"
.SH NAME
GENERIC TETRIS \- Yet Another Tetris Game on X
.SH SYNOPSIS
.B tetris
[-s] [-l <start level>] [-p <prefilled lines>] [-showNext] [-beep]
[-plain] [-u <name>] [-rightrotation] [-scorefile <filename>]
[-display <display>] [-geometry <geometry>] [-iconGeometry <geometry>]
[-{background | bg} <color>] [-{foreground | fg} <color>] [-bigFont
<font>] [-tinyFont <font>] [-xrm <resource specification>]
.SH DESCRIPTION
.LP
This is a game of "\fITetris\fP" for the X Window System.
.LP
The program lists the top 15 players at the end, each player may has at
most three scores. With the option \fB\-s\fP, it shows all the players
and scores.
.LP
Key "\fBj\fP" moves the block left, "\fBl\fP" moves right,
"\fBk\fP" rotates. Left-handed people may want to use "\fBs\fP",
"\fBd\fP", "\fBf\fP" respectively. Or maybe someone wants to try the
arrow keys...  The space bar drops a block quickly. Use "\fBCTRL-L\fP"
key combination to redraw \- in case the program messes somthing up.
.LP
The "\fB+\fP" key speed up or lower down the speed.  To avoid cheating
there is no way to slow down at run-time.  The "\fBb\fP" key toggles
"beep", and "\fBn\fP" toggles "show next".
.LP
If you press "\fBp\fP", the game will be paused, and the window
will be iconified \- just in case your boss suddenly appears at your
door ;o). When the tetris window loses its focus, it also goes to
"paused" mode except it doesn't iconify itself.
.LP
To get high scores, don't use "show next" mode, try starting at a
higher level and try to let "tetris" eat more than one line at a time.
Play \fIfast\fP and see how well you can do in the \fIbonus\fP mode
(what is that? ;-). Warning: some blocks in the bonus mode will rotate
differently than you would expect.
.SH OPTIONS
.PP
.TP 8
.B \-s
.br
This option lets the program only list all the recorded scores then exit.
.TP 8
.B \-l \fIlevel\fP
.br
This option starts the program at the specified level (resource name
"\fIstartLevel\fP").
.TP 8
.B \-p \fIlines\fP
.br
This option indicates that the program should start with some of the lines
being randomly filled (resource name "\fIpreFilledLines\fP").
.TP 8
.B \-showNext
.br
This option tells the program start in "show next" mode (resource name
"\fIshowNext\fP").
.TP 8
.B \-beep
.br
This option turns the sound on (resource name "\fIbeep\fP").
.TP 8
.B \-plain
.br
Uses the old "plain" bitmaps instead of the default 3D ones (resource
name "\fIplain\fP").
.TP 8
.B \-u \fIname\fP
.br
This option sets an alternative player name (resource name
"\fIplayerName\fP").
.TP 8
.B \-rightrotation
Usually the rotation key rotates a piece to the left. Setting this
option makes it rotate to the right (resource name "\fIrightRotation\fP").
.TP 8
.B \-scorefile \fIfilename\fP
Specify an alternative score file (resource name "\fIscoreFile\fP").
.TP 8
.B \-display \fIhost\fP:\fIdpy\fP
.br
This option specifies the X server to contact.
.TP 8
.B \-geometry {+|\-}\fIX\fP{+|\-}\fIY\fP
.br
This option sets the initial position of the tetris window (resource
name "\fIgeometry\fP").
.TP 8
.B \-iconGeometry {+|\-}\fIX\fP{+|\-}\fIY\fP
.br
This option sets the position of the icon (resource name "\fIiconGeometry\fP").
.TP 8
.B \-{background | bg} \fIcolor\fP
.br
This option specifies the background of the tetris window (resource name
"\fIbackground\fP").
.TP 8
.B \-{foreground | fg} \fIcolor\fP
.br
This option specifies the foreground of the tetris window (resource name
"\fIforeground\fP").
.TP 8
.B \-bigFont \fIfont\fP
.br
This option sets the font used in drawing titles and important messages
(resource name "\fIbigFont\fP").
.TP 8
.B \-tinyFont \fIfont\fP
.br
This option sets the font used in drawing status information
(resource name "\fItinyFont\fP").
.TP 8
.B \-xrm <\fIresource specification\fP>
.br
Another method of specifying preferences.
.PP
.SH SEE ALSO
.LP
X(1), xtetris(1), xmris(1)
.SH COPYRIGHTS
.LP
\*R Copyright 1992-95, Q. Alex Zhao, <\fIazhao@cc.gatech.edu\fP>.
.SH BUG REPORTS AND PROGRAM UPDATES
.LP
Send bug reports (or fixes) to the author:
.RS
Q. Alex Zhao,	<\fIazhao@cc.gatech.edu\fP>
.br
College of Computing
.br
Georgia Institute of Technology
.br
Atlanta, GA 30332-0280
.RE
.LP
The most updated source could be found on \fIftp.x.org\fP under
\fI/contrib\fP.
.SH ACKNOWLEDGEMENTS
.LP
Thanks go to people who helped me in improving the program.


tetris.c
/*
# GENERIC X-WINDOW-BASED TETRIS
#
#	tetris.c
#
###
#
#  Copyright (C) 1992 - 95              Q. Alex Zhao, azhao@cc.gatech.edu
#
#			All Rights Reserved
#
#  Permission to use, copy, modify, and distribute this software and
#  its documentation for any purpose and without fee is hereby granted,
#  provided that the above copyright notice appear in all copies and
#  that both that copyright notice and this permission notice appear in
#  supporting documentation, and that the name of the author not be
#  used in advertising or publicity pertaining to distribution of the
#  software without specific, written prior permission.
#
#  This program is distributed in the hope that it will be "playable",
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
#
*/

#include	"tetris.h"
#include	"patchlevel.h"

/*** variables ***/

char            myDisplayName[256];
Display        *display;
int             screen_num;
Visual         *visual;
Bool            useColor = True;
Colormap        colormap;
Window          mainWin, blockWin;
Cursor          theCursor;
XFontStruct    *bigFont, *tinyFont;
unsigned long   fg, bg;
int             rotOffset = 3;

XSizeHints      sizehints, iconsizehints;
XWMHints        wmhints = {
    InputHint | StateHint | IconPixmapHint,
    True,			/* input mode */
    NormalState,		/* normal */
    0,				/* icon pixmap */
    0,				/* icon window */
    0, 0,			/* icon position */
    0				/* icon mask pixmap - not used */
};

int             range = NUM_THINGS;
Bool            bonusMode = False;

char            myHome[FILENAMELEN], scorefile[FILENAMELEN];
int             startlevel = 3, level = 3, prefilled = 0, score = 0, rows = 0;
Bool            showNext = False, beep = False, use3D = True;
score_t         myscore;

static int      opTableEntries = 19;
static XrmOptionDescRec opTable[] = {
    {"-s", "*scoresOnly", XrmoptionNoArg, (caddr_t) "yes"},
    {"-l", "*startLevel", XrmoptionSepArg, (caddr_t) NULL},
    {"-p", "*preFilledLines", XrmoptionSepArg, (caddr_t) NULL},
    {"-showNext", "*showNext", XrmoptionNoArg, (caddr_t) "yes"},
    {"-beep", "*beep", XrmoptionNoArg, (caddr_t) "yes"},
    {"-rightrotation", "*rightRotation", XrmoptionNoArg, (caddr_t) "yes"},
    {"-plain", "*plain", XrmoptionNoArg, (caddr_t) "yes"},
    {"-display", ".display", XrmoptionSepArg, (caddr_t) NULL},
    {"-geometry", "*geometry", XrmoptionSepArg, (caddr_t) NULL},
    {"-iconGeometry", "*iconGeometry", XrmoptionSepArg, (caddr_t) NULL},
    {"-background", "*background", XrmoptionSepArg, (caddr_t) NULL},
    {"-bg", "*background", XrmoptionSepArg, (caddr_t) NULL},
    {"-foreground", "*foreground", XrmoptionSepArg, (caddr_t) NULL},
    {"-fg", "*foreground", XrmoptionSepArg, (caddr_t) NULL},
    {"-bigFont", "*bigFont", XrmoptionSepArg, (caddr_t) NULL},
    {"-tinyFont", "*tinyFont", XrmoptionSepArg, (caddr_t) NULL},
    {"-u", "*playerName", XrmoptionSepArg, (caddr_t) NULL},
    {"-scorefile", "*scoreFile", XrmoptionSepArg, (caddr_t) NULL},
    {"-xrm", NULL, XrmoptionResArg, (caddr_t) NULL}
};
static XrmDatabase cmmDB = (XrmDatabase) NULL, rDB = (XrmDatabase) NULL;

static void parseOpenDisp(int *, char *[]);
static void Usage(char *);
static void getDefaults(void);

/* ------------------------------------------------------------------ */

int
main(int argc, char *argv[])
{
    (void) fprintf(stderr,
		   "                 GENERIC TETRIS %s\n", VERSION);
    (void) fprintf(stderr,
	 "Copyright (C) 1992-95      Q. Alex Zhao, azhao@cc.gatech.edu\n");
    (void) fprintf(stderr,
	 "     GENERIC TETRIS comes with ABSOLUTELY NO WARRANTY.\n\n");

    sizehints.flags = PMinSize | PMaxSize | PPosition | PSize | USSize;
    iconsizehints.flags = PPosition | PSize;
    iconsizehints.width = iconsizehints.height = 48;

    parseOpenDisp(&argc, argv);
    getDefaults();
    inits(argc, argv);
    playing();
    /* never come to here */
    return 0;
}

/* ------------------------------------------------------------------ */

static void
parseOpenDisp(int *argc, char *argv[])
{
    struct passwd  *pw;
    XrmValue        value;
    char           *str_type[20];
    XVisualInfo     vtmp, *vinfo;
    int             n = 1;

    XrmInitialize();

    myDisplayName[0] = '\0';

    XrmParseCommand(&cmmDB, opTable, opTableEntries, "tetris", argc, argv);

    /* check for any arguments left */
    if (*argc != 1) {
	Usage(argv[0]);
    }

    /* only print out the scores? */
    if (XrmGetResource(cmmDB, "tetris.scoresOnly", "Tetris.ScoresOnly",
		       str_type, &value) == True) {

	if (XrmGetResource(cmmDB, "tetris.scoreFile", "Tetris.ScoreFile",
		       str_type, &value) == True) {
	    (void) strncpy(scorefile, value.addr, ZLIM(value.size, FILENAMELEN));
	    scorefile[FILENAMELEN - 1] = '\0';
	} else
	    (void) strncpy(scorefile, SCOREFILE, FILENAMELEN);

	showScores(0);
	exit(0);
    }

    /* get display now */
    if (XrmGetResource(cmmDB, "tetris.display", "Tetris.Display",
		       str_type, &value) == True)
	(void) strncpy(myDisplayName, value.addr, ZLIM(value.size, 255));
    myDisplayName[255] = '\0';

    if (!(display = XOpenDisplay(myDisplayName))) {
	(void) fprintf(stderr, "%s: Can't open display '%s'\n",
		       argv[0], XDisplayName(myDisplayName));
	exit(1);
    }

    screen_num = DefaultScreen(display);
    visual = DefaultVisual(display, screen_num);
    colormap = DefaultColormap(display, screen_num);
    vtmp.visualid = XVisualIDFromVisual(visual);

    if ((vinfo = XGetVisualInfo(display, VisualIDMask, &vtmp, &n)) != NULL) {
	if ((vinfo->class == GrayScale) || (vinfo->class == StaticGray)) {
	    useColor = False;
	}
	XFree((unsigned char *) vinfo);
    } else {
	useColor = False;
    }

    /* setup user information */
    (void) gethostname(myscore.myhost, NAMELEN);

    setpwent();
    pw = getpwuid(getuid());
    if (pw == NULL) {		/* impossible? */
	(void) sprintf(myscore.myname, "%d", getuid());
	myHome[0] = '.';
	myHome[1] = '/';
	myHome[2] = '\0';
    } else {
	(void) strncpy(myscore.myname, pw->pw_name, NAMELEN);
	(void) strncpy(myHome, pw->pw_dir, FILENAMELEN);
    }
    endpwent();
    myscore.myname[NAMELEN - 1] = '\0';
    myscore.myhost[NAMELEN - 1] = '\0';
    myHome[FILENAMELEN - 1] = '\0';
}

/* ------------------------------------------------------------------ */

static void
Usage(char *argv0)
{
    (void) fprintf(stderr,
	"Usage: %s [-s] [-l <starting level>]\n", argv0);
    (void) fprintf(stderr,
	"   [-p <prefilled rows>] [-plain] [-showNext] [-beep] [-u <name>]\n");
    (void) fprintf(stderr,
	"   [-display <display>] [-geometry <geometry>] [-iconGeometry <icon geometry>]\n");
    (void) fprintf(stderr,
	"   [-fg <foreground>] [-bg <background>] [-bigFont <font>] [-tinyFont <font>]\n");
    (void) fprintf(stderr,
	"   [-xrm <resource specifications>]\n");

    exit(1);
}

/* ------------------------------------------------------------------ */

static void
getDefaults(void)
{
    XrmDatabase     homeDB, serverDB, appDB;
    char            filenamebuf[FILENAMELEN];
    char           *filename = &filenamebuf[0];
    char           *env;
    char           *classname = "Tetris";
    char            name[255], geoStr[20], icongeoStr[20];
    XrmValue        value;
    char           *str_type[20];
    int             x, y;
    unsigned int    w, h;
    long            flags;

    (void) strcpy(name, "/usr/lib/X11/app-defaults/");
    (void) strcat(name, classname);

    /* Get application defaults file, if any */
    appDB = XrmGetFileDatabase(name);
    (void) XrmMergeDatabases(appDB, &rDB);

    if (XResourceManagerString(display) != NULL) {
	serverDB = XrmGetStringDatabase(XResourceManagerString(display));
    } else {
	(void) strcpy(filename, myHome);
	(void) strcat(filename, "/.Xdefaults");
	serverDB = XrmGetFileDatabase(filename);
    }
    XrmMergeDatabases(serverDB, &rDB);

    if ((env = getenv("XENVIRONMENT")) == NULL) {
	int             len;

	env = &filenamebuf[0];
	(void) strcpy(env, myHome);
	len = strlen(env);
	env[len++] = '/';
	(void) gethostname(env + len, FILENAMELEN - len);
    }
    homeDB = XrmGetFileDatabase(env);
    XrmMergeDatabases(homeDB, &rDB);

    XrmMergeDatabases(cmmDB, &rDB);

    /* starting levels */

    if (XrmGetResource(rDB, "tetris.startLevel", "Tetris.StartLevel",
		       str_type, &value) == True) {
	if ((sscanf(value.addr, "%d", &level) <= 0) ||
		(level < 0) || (level >= NUM_LEVELS)) {
	    (void) fprintf(stderr, "Tetris: Invalid level.\n");
	    exit(1);
	}

	startlevel = level;
    }

    /* prefilled lines */

    if (XrmGetResource(rDB, "tetris.preFilledLines", "Tetris.PreFilledLines",
		       str_type, &value) == True) {
	if ((sscanf(value.addr, "%d", &prefilled) <= 0) ||
		(prefilled < 0) || (prefilled >= ROWS - THINGSIZE)) {
	    (void) fprintf(stderr, "Tetris: Invalid prefilled lines.\n");
	    exit(1);
	}
    }

    /* show next */

    if (XrmGetResource(rDB, "tetris.showNext", "Tetris.ShowNext",
			str_type, &value) == True) {
	showNext = True;
    }

    /* bitmaps */

    if (XrmGetResource(rDB, "tetris.plain", "Tetris.Plain",
			str_type, &value) == True) {
	use3D = False;
    }

    /* beep */

    if (XrmGetResource(rDB, "tetris.beep", "Tetris.Beep",
			str_type, &value) == True) {
	beep = True;
    }

    /* rotation */

    if (XrmGetResource(rDB, "tetris.rightRotation", "Tetris.RightRotation",
			str_type, &value) == True) {
	rotOffset = 1;
    }

    /*** get foreground/background colors ***/

    if (XrmGetResource(rDB, "tetris.foreground", "Tetris.Foreground",
		       str_type, &value) == True) {
	(void) strncpy(name, value.addr, ZLIM(value.size, 255));
	name[254] = '\0';
	fg = getColor(name);
    } else
	fg = BlackPixel(display, screen_num);

    if (XrmGetResource(rDB, "tetris.background", "Tetris.Background",
		       str_type, &value) == True) {
	(void) strncpy(name, value.addr, ZLIM(value.size, 255));
	name[254] = '\0';
	bg = getColor(name);
    } else
	bg = WhitePixel(display, screen_num);

    if (bg == fg) {
	bg = WhitePixel(display, screen_num);
	fg = BlackPixel(display, screen_num);
    }

    /*** get geometry info ***/

    if (XrmGetResource(rDB, "tetris.geometry", "Tetris.Geometry",
		       str_type, &value) == True) {
	(void) strncpy(geoStr, value.addr, ZLIM(value.size, 20));
	geoStr[19] = '\0';
    } else {
	geoStr[0] = '\0';
    }

    flags = XParseGeometry(geoStr, &x, &y, &w, &h);
    if ((WidthValue | HeightValue) & flags)
	Usage("tetris");

    if (XValue & flags) {
	if (XNegative & flags)
	    x = DisplayWidth(display, screen_num) + x - sizehints.width;
	sizehints.x = x;
    }
    if (YValue & flags) {
	if (YNegative & flags)
	    y = DisplayHeight(display, screen_num) + y - sizehints.height;
	sizehints.y = y;
    }

    /*** get icon geometry info ***/

    if (XrmGetResource(rDB, "tetris.iconGeometry", "Tetris.IconGeometry",
		       str_type, &value) == True) {
	(void) strncpy(icongeoStr, value.addr, ZLIM(value.size, 20));
	icongeoStr[19] = '\0';
    } else {
	icongeoStr[0] = '\0';
    }

    flags = XParseGeometry(icongeoStr, &x, &y, &w, &h);
    if ((WidthValue | HeightValue) & flags)
	Usage("tetris");

    if (XValue & flags) {
	if (XNegative & flags)
	    x = DisplayWidth(display, screen_num) + x - iconsizehints.width;
	wmhints.flags |= IconPositionHint;
	wmhints.icon_x = x;
	iconsizehints.x = x;
    }
    if (YValue & flags) {
	if (YNegative & flags)
	    y = DisplayHeight(display, screen_num) + y - iconsizehints.height;
	wmhints.flags |= IconPositionHint;
	wmhints.icon_y = y;
	iconsizehints.y = y;
    }

    /*** get fonts ***/

    if (XrmGetResource(rDB, "tetris.bigFont", "Tetris.BigFont",
		       str_type, &value) == True) {
	(void) strncpy(name, value.addr, ZLIM(value.size, 255));
	name[254] = '\0';
    } else {
	(void) strcpy(name, BIGFONT);
    }
    if ((bigFont = XLoadQueryFont(display, name)) == NULL) {
	(void) fprintf(stderr, "Tetris: can't open font '%s'.\n", name);
	exit(1);
    }
    if (XrmGetResource(rDB, "tetris.tinyFont", "Tetris.TinyFont",
		       str_type, &value) == True) {
	(void) strncpy(name, value.addr, ZLIM(value.size, 255));
	name[254] = '\0';
    } else {
	(void) strcpy(name, TINYFONT);
    }
    if ((tinyFont = XLoadQueryFont(display, name)) == NULL) {
	(void) fprintf(stderr, "Tetris: can't open font '%s'.\n", name);
	exit(1);
    }

    /* player name */

    if (XrmGetResource(rDB, "tetris.playerName", "Tetris.PlayerName",
		       str_type, &value) == True) {
	(void) strncpy(myscore.myname, value.addr, ZLIM(value.size, NAMELEN));
	myscore.myname[NAMELEN - 1] = '\0';
    }

    /* score file */

    if (XrmGetResource(rDB, "tetris.scoreFile", "Tetris.ScoreFile",
		       str_type, &value) == True) {
	(void) strncpy(scorefile, value.addr, ZLIM(value.size, FILENAMELEN));
	scorefile[FILENAMELEN - 1] = '\0';
    } else
	(void) strncpy(scorefile, SCOREFILE, FILENAMELEN);

    /*
     * clean up
     */
    XrmDestroyDatabase(rDB);
}

/* ------------------------------------------------------------------ */
